numpy/core/src/multiarray/conversion_utils.c

```diff
@@ -110,42 +130,77 @@ PyArray_IntpConverter(PyObject *obj, PyArray_Dims *seq)
         return NPY_SUCCEED;
     }
 
-    len = PySequence_Size(obj);
-    if (len == -1) {
-        /* Check to see if it is an integer number */
-        if (PyNumber_Check(obj)) {
-            /*
-             * After the deprecation the PyNumber_Check could be replaced
-             * by PyIndex_Check.
-             * FIXME 1.9 ?
-             */
-            len = 1;
+    PyObject *seq_obj = NULL;
+
+    /*
+    * If obj is a scalar we skip all the useless computations and jump to
+    * intp_from_scalar as soon as possible.
+    */
+    if (!PyLong_CheckExact(obj) && PySequence_Check(obj)) {
+        seq_obj = PySequence_Fast(obj,
+               "Expected an integer or an iterable/sequence of integers.");
+        if (seq_obj == NULL) {
+            // we ignore this error since it is managed below
+            PyErr_Clear();
         }
     }
-    if (len < 0) {
-        PyErr_SetString(PyExc_TypeError,
-                "expected sequence object with len >= 0 or a single integer");
-        return NPY_FAIL;
-    }
-    if (len > NPY_MAXDIMS) {
-        PyErr_Format(PyExc_ValueError, "maximum supported dimension for an ndarray is %d"
-                     ", found %d", NPY_MAXDIMS, len);
-        return NPY_FAIL;
-    }
-    if (len > 0) {
-        seq->ptr = npy_alloc_cache_dim(len);
+
+    if (seq_obj == NULL) {
+        /*
+        * obj *might* be a scalar (if intp_from_scalar does not fail, at the
+        * moment no check have been performed to verify this hypothesis).
+        */
+
+        seq->ptr = npy_alloc_cache_dim(1);
         if (seq->ptr == NULL) {
             PyErr_NoMemory();
             return NPY_FAIL;
         }
+        seq->len = 1;
+
+        seq->ptr[0] = intp_from_scalar(obj);
+        /*
+        * Check whether the value set by intp_from_scalar is different than
+        * the value returned, which means that an error occurred.
+        */
+        if (PyErr_Occurred() != NULL) {
+            npy_free_cache_dim_obj(*seq);
+            seq->ptr = NULL;
```

Comment:
Above we have this error:
```
        PyErr_SetString(PyExc_TypeError,
                 "expected sequence object with len >= 0 or a single integer");
```
which is the "all else failed" path.  I would rmeove the `len >= 0` it doesn't make much sense.  But otherwise, it seems sensible to preserve it.  Could make it `expected a sequence of integers or a single integer`.

To be clear: This is basically the "all else failed" path here, since we now already:
* Checked whether it is a sequence (it may not contain integers, but we do not reach if it is a sequence).
* Checked whether it is an integer.

Reply:
Fixed in c545c3dbca2f0d460d7a593c42654d4779c95dd3, is this what you expected?

Reply:
Had to fix this a little bit in 9add7a7a025c454d68db9928da2e493137d959d9 since one of the tests which checks the returned exception when an overflow occurs expected a `ValueError` instead of a `TypeError`.